python/lsst/daf/base/propertyContainer/propertyContainerContinued.py

```diff
@@ -1056,3 +1103,49 @@ def __reduce__(self):
         # object.__new__(PropertyList, *args) which bypasses
         # the pybind11 memory allocation step.
         return (_makePropertyList, (getPropertyListState(self),))
+
+    def get_dict(self, key: str) -> NestedMetadataDict:
+        """Return a possibly-hierarchical nested `dict`.
+
+        This implements the `lsst.pipe.base.GetDictMetadata` protocol for
+        consistency with `lsst.pipe.base.TaskMetadata` and `PropertySet`.
+
+        Parameters
+        ----------
+        key : `str`
+            String key associated with the mapping.
+
+        Returns
+        -------
+        value : `~collections.abc.Mapping`
+            Possibly-nested mapping, with `str` keys and values that are `int`,
+            `float`, `str`, `bool`, or another `dict` with the same key and
+            value types.  Will be empty if ``key`` does not exist.
+        """
+        result: NestedMetadataDict = {}
+        name: str
+        for name in self.getOrderedNames():
+            levels = name.split(".")
+            if levels[0] == key:
```

Comment:
I think in theory you can break out the loop if you've matched in a previous loop but then next time around the key does not match (since the ordered names means that `key` is guaranteed not to appear again). Not sure if it matters unless the FITS header is huge.

This does make me think that when we read HIERARCH headers we should have been replacing the spaces with `.`.

Also, I was trying to see what happened if you added a PropertySet to a PropertyList twice. It's not pretty.

```python
>>> from lsst.daf.base import PropertyList, PropertySet
>>> pl = PropertyList()
>>> pl["A"] = "B"
>>> ps = PropertySet.from_mapping({"B": "C", "D": "E"})
>>> ps2 = PropertySet.from_mapping({"X": "Y", "Z": "W"})
>>> pl.set("AA", ps)
>>> pl.set("AA", ps2)
>>> print(pl)
A = "B"
AA.D = "E"
AA.B = "C"
AA.Z = "W"
AA.X = "Y"

>>> pl.add("AA", ps)
>>> print(pl)
zsh: segmentation fault  python
```

Reply:
Breaking out of the loop because we expect the call to `set_dict` to be atomic also makes it more directly thread-unsafe (rather than merely thread-unsafe because it's probably calling thread-unsafe things).  I'm inclined to leave it as it is, since we're doing an O(N) operation in getting the full set of keys to iterate over already.

This is all making me _slightly_ interested in writing a replacement for `PropertyList` as our in-memory thing that maps to FITS headers and _really_ uninterested in fixing `PropertyList` itself.  I think right now I just have to hold my nose and get back to other things.